### PR TITLE
Update life_of_a_prow_job.md

### PR DESCRIPTION
**Pull request by nicoaragon**
Refreshing documentation.

/assign BenTheElder

### CODE REVIEW

**Comment by BenTheElder** [comment]
/ok-to-test
/assign @cjwagner
/unassign
🙃

File: prow/life_of_a_prow_job.md [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
+The "hook" service routes traffic to the "hook" application. [This deployment resource](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/cluster/hook_deployment.yaml) defines the pods that "hook" is comprised of.
+
+The pods for "hook" run [the "Hook" executable](https://github.com/kubernetes/test-infra/tree/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/cmd/hook). "hook" listens for incoming http requests and translates them to "GitHub event objects". Afterwards, "hook" broadcasts these events to Prow Plugins. In the case of my `/test all` comment, "hook" builds an ["Issue Comment Event"](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/github/types.go#L116-L121).
+
+Prow plugins receive 2 objects: 1) a GitHub event object, and 2) a ["plugin client object"](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/plugins.go#L64-L70). Plugin client objects contain 3-clients: a GitHub client, a Kubernetes client, and a logging client. This is what "plugin client objects" look like:
```
Comment (alvaroaleman) [inline prow/life_of_a_prow_job.md]: Is the hyphen in `3-clients` intended?

File: prow/life_of_a_prow_job.md [hunk 2]
```diff
@@ -15,7 +19,13 @@ type PluginClient struct {
...
+
+"hook" delivers an event that represents my `/test all` comment to the [Trigger plugin](https://github.com/kubernetes/test-infra/tree/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/trigger). The trigger plugin uses a function called [`handleGenericCommentEvent` (implementec by `handleGenericComment`)](https://github.com/kubernetes/test-infra/blob/99b91b56b097e39d70cb1ae82c0b1cb57d98ac48/prow/plugins/trigger/generic-comment.go#L32) to process `/test all`.
+
+`handleGenericComment` verifies that it is "ok to test", and then stamps a label that reads "ok to test" on the issue. Finally, `handleGenericComment` determines *presubmit* jobs to run. For `/test all`, it will consider jobs that need no people interaction. The list of jobs supplied by the `Config` object, in the `PluginClient` object, will be used to find suitable jobs.
```
Comment (alvaroaleman) [inline prow/life_of_a_prow_job.md]: Should we really describe this at that level of detail here? What about a more high level `The trigger plugin checks if the Pull Request is deemed safe to test by checking if the author is an org member or the PR has an "ok-to-test" label?`

Also this isn't quite correct, as PRs authored by org members do not need the `ok-to-test` label

File: prow/life_of_a_prow_job.md [hunk 3]
```diff
@@ -15,7 +19,13 @@ type PluginClient struct {
...
+
+I requested that some tests run, so the trigger plugin talks to the Kubernetes API server and creates a [ProwJob Custom Resource](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/kube/prowjob.go#L50-L83) with the information from the issue comment. At this point, `cmd/hook`'s role is done. It goes on and handles the next webhooks.
```
Comment (alvaroaleman) [inline prow/life_of_a_prow_job.md]: The whole paragraph started out with `I comment /test all` or not? That means `trigger` will check which presubmits are mandatory for your PR and run those, so the `I requested that some tests run` doesn't sound accurate to me :thinking:

File: prow/life_of_a_prow_job.md (hunk 1 above)
Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: fixed :-)

File: prow/life_of_a_prow_job.md (hunk 2 above)
Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: I agree. This is fixed.

File: prow/life_of_a_prow_job.md (hunk 3 above)
Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: fixed :-)

File: prow/life_of_a_prow_job.md [hunk 4]
```diff
@@ -15,9 +19,15 @@ type PluginClient struct {
...
+
+"hook" delivers an event that represents my `/test all` comment to the [Trigger plugin](https://github.com/kubernetes/test-infra/tree/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/trigger). The Trigger plugin validates the PR before running tests. One such validation is, for instance, that the author is a member of the organization or that the PR is labeled `ok-to-test`. The function called [`handleGenericCommentEvent` (implementec by `handleGenericComment`)](https://github.com/kubernetes/test-infra/blob/99b91b56b097e39d70cb1ae82c0b1cb57d98ac48/prow/plugins/trigger/generic-comment.go#L32) describes Trigger's logic.
+
+Finally, `handleGenericComment` determines *presubmit* jobs to run. For `/test all`, it will consider jobs that need no people interaction. The list of jobs supplied by the `Config` object, in the `PluginClient` object, will be used to find suitable jobs.
```
Comment (alvaroaleman) [inline prow/life_of_a_prow_job.md]: There is no such concept as "jobs that need no people interaction". `/test all` triggers all jobs that are considered mandatory, based off if a job is optional or has a `run_if_changed` setting

Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: I agree!

Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: comments addressed! Thank you for the feedback!

**Comment by nicoaragon** [comment]
Thank you! How do I merge the changes?

**Comment by alvaroaleman** [comment]
> Thank you! How do I merge the changes?

By having a codeowner approve it, see https://github.com/kubernetes/test-infra/blob/master/CONTRIBUTING.md#workflow

/assign @stevekuznetsov

**Comment by alvaroaleman** [comment]
@nicoaragon Also, please squash your commits together so there is only one

**Comment by nicoaragon** [comment]
Thank you, @alvaroaleman ! I think I did it.

If that did not work is there an option to squash as you merge?

**Comment by alvaroaleman** [comment]
> If that did not work is there an option to squash as you merge?

Unfortunately not. Here is an explanation on how you can do it: https://github.com/wprig/wprig/wiki/How-to-squash-commits

**Comment by nicoaragon** [comment]
@alvaroaleman I was silly with console commands. Squashed the commits.

**Comment by nicoaragon** [comment]
Thank you for the link! :-)

**Comment by nicoaragon** [comment]
What are next steps?

File: prow/life_of_a_prow_job.md [hunk 5]
```diff
@@ -1,10 +1,14 @@
 # Life of a Prow Job
 
-I comment `/test all` on a PR. GitHub creates a JSON object representing that action and sends a webhook to prow. [Here](https://github.com/kubernetes/test-infra/tree/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/cmd/phony/examples) are some examples of webhook payloads.
```
[the remark this answers is not in the document]
Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: Good point. Added a link for those.

[the remark this answers is not in the document]
Comment (nicoaragon) [reply prow/life_of_a_prow_job.md]: I addressed remaining comments below. Thank you!

**Comment by nicoaragon** [comment]
thank you, @cjwagner ! I incorporated changes based on your feedback. What do you think?

**Comment by nicoaragon** [comment]
Replaced by [/test-infra/pull/14040](https://github.com/kubernetes/test-infra/pull/14040)

**Pull request closed without merging**